cmake/optional.cmake

```diff
@@ -0,0 +1,234 @@
+)
+option(ALLOW_INCOMPATIBLE_BDB "Allow using a Berkeley DB (BDB) version other than 4.8" OFF)
+
+set(WITH_NATPMP "auto" CACHE STRING "Enable NAT-PMP ([auto], yes, no). \"auto\" means \"yes\" if libnatpmp is found")
+option(ENABLE_NATPMP_DEFAULT "If NAT-PMP is enabled, turn it on at startup" OFF)
+
+set(WITH_MINIUPNPC "auto" CACHE STRING "Enable UPNP ([auto], yes, no). \"auto\" means \"yes\" if libminiupnpc is found")
+option(ENABLE_UPNP_DEFAULT "If UPNP is enabled, turn it on at startup" OFF)
+
+set(WITH_ZMQ "auto" CACHE STRING "Enable ZMQ notifications ([auto], yes, no). \"auto\" means \"yes\" if libzmq is found")
+
+set(WITH_USDT "auto" CACHE STRING "Enable tracepoints for Userspace, Statically Defined Tracing ([auto], yes, no). \"auto\" means \"yes\" if sys/sdt.h is found")
+
+set(WITH_QRENCODE "auto" CACHE STRING "Enable QR code support ([auto], yes, no). \"auto\" means \"yes\" if libqrencode is found")
+
+set(WITH_SECCOMP "auto" CACHE STRING "Enable experimental syscall sandbox feature (-sandbox) ([auto], yes, no). \"auto\" means \"yes\" if seccomp-bpf is found under Linux x86_64")
+
+set(WITH_EXTERNAL_SIGNER "auto" CACHE STRING "Enable external signer support ([auto], yes, no). \"auto\" means \"yes\" if Boost.Process is found")
+
+set(OPTION_VALUES auto yes no)
+foreach(option USE_CCACHE WITH_SQLITE WITH_BDB WITH_NATPMP WITH_MINIUPNPC WITH_ZMQ WITH_USDT WITH_QRENCODE WITH_SECCOMP)
+  if(NOT ${option} IN_LIST OPTION_VALUES)
+    message(FATAL_ERROR "${option} value is \"${${option}}\", but must be one of \"auto\", \"yes\" or \"no\".")
+  endif()
+endforeach()
+
+if(NOT USE_CCACHE STREQUAL "no")
+  find_program(CCACHE ccache)
+  if(CCACHE)
+    set(CMAKE_C_COMPILER_LAUNCHER ${CCACHE})
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE})
+  elseif(USE_CCACHE STREQUAL "yes")
+    message(FATAL_ERROR "ccache requested, but not found.")
+  endif()
+endif()
+
+if(ENABLE_WALLET)
+  if(NOT WITH_SQLITE STREQUAL "no")
+    cross_pkg_check_modules(sqlite sqlite3>=3.7.17 IMPORTED_TARGET)
+    if(sqlite_FOUND)
+      set(WITH_SQLITE yes)
+      set(USE_SQLITE ON)
+    else()
+      if(WITH_SQLITE STREQUAL "yes")
+        message(FATAL_ERROR "SQLite requested, but not found.")
+      endif()
+      set(WITH_SQLITE no)
+    endif()
+  endif()
+
+  if(NOT WITH_BDB STREQUAL "no")
+    find_package(BerkeleyDB 4.8)
+    if(BerkeleyDB_FOUND)
+      set(WITH_BDB yes)
+      set(USE_BDB ON)
+      if(NOT BerkeleyDB_VERSION VERSION_EQUAL 4.8)
+        message(WARNING "Found Berkeley DB (BDB) other than 4.8.")
+        if(ALLOW_INCOMPATIBLE_BDB)
+          message(WARNING "BDB (legacy) wallets opened by this build will not be portable!")
+        else()
+          message(WARNING "BDB (legacy) wallets opened by this build would not be portable!\n"
+                          "If this is intended, pass \"-DALLOW_INCOMPATIBLE_BDB=ON\".\n"
+                          "Passing \"-DWITH_BDB=no\" will suppress this warning.\n")
```

Comment:
Is the only difference in the text of the warning message, if `ALLOW_INCOMPATIBLE_BDB` is defined or not? I would expect the build to fail if bdb 5.3 is found and `ALLOW_INCOMPATIBLE_BDB` is not defined. Also, the warnings look a bit too alarming in the cmake output, especially given that I have specified `-DALLOW_INCOMPATIBLE_BDB=ON`. Consider the below or feel free to ignore:

```diff
--- i/cmake/optional.cmake
+++ w/cmake/optional.cmake
@@ -69,19 +69,19 @@ if(ENABLE_WALLET)
   if(NOT WITH_BDB STREQUAL "no")
     find_package(BerkeleyDB 4.8)
     if(BerkeleyDB_FOUND)
       set(WITH_BDB yes)
       set(USE_BDB ON)
       if(NOT BerkeleyDB_VERSION VERSION_EQUAL 4.8)
-        message(WARNING "Found Berkeley DB (BDB) other than 4.8.")
+        message(STATUS "Found Berkeley DB (BDB) other than 4.8.")
         if(ALLOW_INCOMPATIBLE_BDB)
-          message(WARNING "BDB (legacy) wallets opened by this build will not be portable!")
+          message(STATUS "BDB (legacy) wallets opened by this build will not be portable!")
         else()
-          message(WARNING "BDB (legacy) wallets opened by this build would not be portable!\n"
+          message(FATAL_ERROR "BDB (legacy) wallets opened by this build would not be portable!\n"
                           "If this is intended, pass \"-DALLOW_INCOMPATIBLE_BDB=ON\".\n"
-                          "Passing \"-DWITH_BDB=no\" will suppress this warning.\n")
+                          "Passing \"-DWITH_BDB=no\" will suppress this check.\n")
         endif()
       endif()
```

Reply:
There is a rough consensus among build system people here that the new CMake-based system should follow the current Autotools-based one as close as possible.

These lines follow https://github.com/bitcoin/bitcoin/blob/fa5c224d444802dabec5841009e029b9754c92f1/build-aux/m4/bitcoin_find_bdb48.m4#L56-L66

which was recently updated in #23168.